File: tests/scripts/requestsTests/getTests.cpp
#include "utils.cpp"

void	getTest(const std::string &url, const std::string &tabs)
{

	testServers("", url, tabs, "GET");
}

void	getTests(void)
{
	printHeader("\t Classic Case");
	getTest( "/get/main.html", "\t\t\t\t\t");
	getTest( "/get/fake/main.cpp", "\t\t\t\t\t");
	getTest( "/get/fake/../main.html", "\t\t\t\t");
	getTest( "/get/fake/../../../../../../../../main.html", "\t\t");
	getTest( "/get/../main.html", "\t\t\t\t\t");

	printHeader("\t Index Case");
	getTest( "/get/srcs/", "\t\t\t\t\t\t");

	printHeader("\t Redirection Case");
	getTest( "/get/srcs", "\t\t\t\t\t\t");

	printHeader("\t 403 Case");
	getTest( "/get/fake/", "\t\t\t\t\t\t");
	getTest( "/get/nonono/", "\t\t\t\t\t\t");

	printHeader("\t 405 Case");
	getTest( "/get/405/", "\t\t\t\t\t\t");

	printHeader("\t 404 Case");
	getTest( "/unitTest/uplo/", "\t\t\t\t\t");
	getTest( "/gknrk", "\t\t\t\t\t\t");
	getTest( "/bin/", "\t\t\t\t\t\t\t");

	printHeader("\t Auto Index");
	getTest( "/get/auto/", "\t\t\t\t\t\t");
	getTest( "/get/auto2/", "\t\t\t\t\t\t");

	printHeader("\t Folder With File Name");
	getTest( "/get/truc.txt/", "\t\t\t\t\t");

	printHeader("\t Index Specials");
	getTest( "/get/indexIsFolder/", "\t\t\t\t\t");
	getTest( "/get/indexIsSymlink/", "\t\t\t\t\t");
	getTest( "/get/indexIsFolderAndFile/", "\t\t\t\t");
	getTest( "/get/indexIsSymlinkAndFile/", "\t\t\t\t");

	printHeader("\t Fix Url Errors");
	getTest( "/get//main.html", "\t\t\t\t\t");
	getTest( "/get/endWith./index.html", "\t\t\t\t");
}

int	main(void)
{
	if (std::system("curl -I localhost:8181 > /dev/null 2>&1") != 0)
	{
		std::cout << "Nginx server not running, exiting" << std::endl;
		return (EXIT_FAILURE);
	}
	if (std::system("curl -I localhost:8080 > /dev/null 2>&1") != 0)
	{
		std::cout << "Webserv server not running, exiting" << std::endl;
		return (EXIT_FAILURE);
	}

	if (std::system("mkdir -p ../../../code/unitTest && cd ../../../code/unitTest && ../../tests/scripts/requestsTests/initGetTest.sh") != 0  // For our server
		|| std::system("cd ../../website && ../scripts/requestsTests/initGetTest.sh") != 0)  // For nginx
	{
		std::cout << "Error executing the get init scripts" << std::endl;
		return (EXIT_FAILURE);
	}

	getTests();


	if (std::system("cd ../../../code/unitTest && ../../tests/scripts/requestsTests/cleanGetTest.sh") != 0 // For our server
		|| std::system("cd ../../website && ../scripts/requestsTests/cleanGetTest.sh") != 0) // For nginx
	{
		std::cout << "Error executing the get cleaning scripts" << std::endl;
		return (EXIT_FAILURE);
	}
	return (EXIT_SUCCESS);
}

File: tests/scripts/requestsTests/getTests.mjs
import { compareGoodRequests } from "./compareRequests.mjs"
import { verifyServersAreRunning, exec, printHeader, COLOR_GREEN, COLOR_RED, COLOR_RESET } from "./utils.mjs"

let	failedTests = [];

async function	runGoodGetTest(target, expectedHeaders)
{
	const	result = await compareGoodRequests(target, "GET", null, {}, expectedHeaders);
	if (result == true)
		console.log(COLOR_GREEN + "[OK] " + COLOR_RESET);
	else
	{
		console.log(COLOR_RED + "[KO] " + COLOR_RESET);
		failedTests.push(target);
	}
	return (result);
}

async function runTests()
{
	printHeader("Classic Case");
	await runGoodGetTest("/get/main.html");
	await runGoodGetTest("/get/fake/main.cpp");
	await runGoodGetTest("/get/fake/../main.html");
	await runGoodGetTest("/get/fake/../../../../../../../../main.html");
	await runGoodGetTest("/get/../main.html");

	printHeader("Index Case");
	await runGoodGetTest("/get/srcs/");

	printHeader("Redirection Case");
	await runGoodGetTest("/get/srcs");

	printHeader("403 Case");
	await runGoodGetTest("/get/fake/");
	await runGoodGetTest("/get/nonono/");

	printHeader("405 Case");
	await runGoodGetTest("/get/405/");

	printHeader("404 Case");
	await runGoodGetTest("/unitTest/uplo/");
	await runGoodGetTest("/gknrk");
	await runGoodGetTest("/bin/");

	printHeader("Auto Index");
	await runGoodGetTest("/get/auto/");
	await runGoodGetTest("/get/auto2/");

	printHeader("Folder With File Name");
	await runGoodGetTest("/get/truc.txt/");

	printHeader("Index Specials");
	await runGoodGetTest("/get/indexIsFolder/");
	await runGoodGetTest("/get/indexIsSymlink/");
	await runGoodGetTest("/get/indexIsFolderAndFile/");
	await runGoodGetTest("/get/indexIsSymlinkAndFile/");

	printHeader("Fix Url Errors");
	await runGoodGetTest("/get//main.html");
	await runGoodGetTest("/get/endWith./index.html");

	printHeader("Add_header Tests Not Always With Error")
	await runGoodGetTest("/add_header/get/srcs/");

	printHeader("Add_header Tests Always With Error")
	await runGoodGetTest("/add_header/always/get/srcs/", {"pragma":"truc"});

	printHeader("Add_header Tests Not Always Without Error")
	await runGoodGetTest("/add_header/get/srcs/main.cpp", {"accept":"text/html"});

	printHeader("Add_header Tests Always Without Error")
	await runGoodGetTest("/add_header/always/get/srcs/main.cpp", {"pragma":"truc"});

	printHeader("Same but with query string");
	await runGoodGetTest("/get/main.html?truc=var");
	await runGoodGetTest("/get/fake/main.cpp?truc=var");
	await runGoodGetTest("/get/fake/../main.html?truc=var");
	await runGoodGetTest("/get/fake/../../../../../../../../main.html?truc=var");
	await runGoodGetTest("/get/../main.html?truc=var");
	await runGoodGetTest("/get/srcs/?truc=var");
	await runGoodGetTest("/get/srcs?truc=var");
	await runGoodGetTest("/get/fake/?truc=var");
	await runGoodGetTest("/get/nonono/?truc=var");
	await runGoodGetTest("/get/405/?truc=var");
	await runGoodGetTest("/unitTest/uplo/?truc=var");
	await runGoodGetTest("/gknrk?truc=var");
	await runGoodGetTest("/get/auto2/?truc=var");
	await runGoodGetTest("/get/truc.txt/?truc=var");
	await runGoodGetTest("/get/indexIsFolder/?truc=var");
	await runGoodGetTest("/get/indexIsSymlink/?truc=var");
	await runGoodGetTest("/get/indexIsFolderAndFile/?truc=var");
	await runGoodGetTest("/get/indexIsSymlinkAndFile/?truc=var");
	await runGoodGetTest("/get//main.html?truc=var");
	await runGoodGetTest("/get/endWith./index.html?truc=var");

	if (failedTests.length == 0)
		printHeader("Everything Done : " + COLOR_GREEN + "[OK] " + COLOR_RESET);
	else
	{
		printHeader("Everything Done : " + COLOR_RED + "[KO] " + COLOR_RESET);
		console.table(failedTests);
	}
}

async function	run()
{
	verifyServersAreRunning();

	exec("cd ../../webserv && ../scripts/requestsTests/initGetTest.sh"); // setup webserv
	exec("cp -pr ../../webserv/* ../../nginx/"); // setup nginx

	await runTests();

	exec("cd ../../webserv && ../scripts/requestsTests/cleanGetTest.sh"); // clean webserv
	exec("rm -rf ../../nginx/*"); // clean nginx
}

run();
